airflow/jobs/scheduler_job.py

```diff
@@ -1749,62 +1403,353 @@ def _run_scheduler_loop(self) -> None:
+        """
+        Make scheduling decisions about an individual dag run
+
+        ``currently_active_runs`` is passed in so that a batch query can be
+        used to ask this for all dag runs in the batch, to avoid an n+1 query.
+
+        :param dag_run: The DagRun to schedule
+        :param currently_active_runs: Number of currently active runs of this DAG
+        :return: Number of tasks scheduled
+        """
+        dag = dag_run.dag = self.dagbag.get_dag(dag_run.dag_id, session=session)
+
+        if not dag:
+            self.log.error(
+                "Couldn't find dag %s in DagBag/DB!", dag_run.dag_id
+            )
+            return 0
+
+        if (
+            dag_run.start_date and dag.dagrun_timeout and
+            dag_run.start_date < timezone.utcnow() - dag.dagrun_timeout
+        ):
+            dag_run.state = State.FAILED
+            dag_run.end_date = timezone.utcnow()
+            self.log.info("Run %s of %s has timed-out", dag_run.run_id, dag_run.dag_id)
+            session.flush()
+
+            # Work out if we should allow creating a new DagRun now?
+            self._update_dag_next_dagruns([session.query(DagModel).get(dag_run.dag_id)], session)
+
+            callback_to_execute = DagCallbackRequest(
+                full_filepath=dag.fileloc,
+                dag_id=dag.dag_id,
+                execution_date=dag_run.execution_date,
+                is_failure_callback=True,
+                msg='timed_out'
+            )
+
+            # Send SLA & DAG Success/Failure Callbacks to be executed
+            self._send_dag_callbacks_to_processor(dag_run, callback_to_execute)
+
+            return 0
+
+        if dag_run.execution_date > timezone.utcnow() and not dag.allow_future_exec_dates:
+            self.log.error(
+                "Execution date is in future: %s",
+                dag_run.execution_date
+            )
+            return 0
+
+        if dag.max_active_runs:
+            if currently_active_runs >= dag.max_active_runs:
+                self.log.info(
+                    "DAG %s already has %d active runs, not queuing any more tasks",
+                    dag.dag_id,
+                    currently_active_runs,
+                )
+                return 0
+
+        self._verify_integrity_if_dag_changed(dag_run=dag_run, session=session)
+        # TODO[HA]: Rename update_state -> schedule_dag_run, ?? something else?
+        schedulable_tis, callback_to_run = dag_run.update_state(session=session, execute_callbacks=False)
+
+        self._send_dag_callbacks_to_processor(dag_run, callback_to_run)
+
+        # Get list of TIs that do not need to executed, these are
+        # tasks using DummyOperator and without on_execute_callback / on_success_callback
+        dummy_tis = [
+            ti for ti in schedulable_tis
+            if
+            (
+                ti.task.task_type == "DummyOperator"
```

Comment:
[Previously](https://github.com/apache/airflow/blob/eb5fea7b647e0f4b2338f86eeff6a90e1d6043bc/airflow/jobs/scheduler_job.py#L985), I used `isinstance` to handle still other operators that have no action e.g. [ExternalTaskMarker](https://github.com/apache/airflow/blob/eb5fea7b647e0f4b2338f86eeff6a90e1d6043bc/airflow/sensors/external_task_sensor.py#L225). Can you correct or add a TODO note about this limitation?

Reply:
Oh yes. We can't use `isinstance` anymore due to the serialization, unfortunately. Will todo/add an issue shortly.

Reply:
I know. For now, we can add a list of all the classes from the core that inherit from DummyOperator. This should be sufficient in most cases.